#Part C
#I aimed to construct a pylogenetic relationship among Staphylococcus species. I first retrieved sequences from NCBI. I used 16S gene as it is the most common method used for Staphylococcus identification Then I create a datafram with column names that includes my sequences and their titles. After checking my sequences I aligned them using Muscle function so that I can compute a distance matrix based on k-mers and finally using hclust function I clustered them and drew a plot.
library(tidyverse)
library(stringi)
library(ape)
library(Biostrings)
library(muscle)
library(msa)
library(DECIPHER)
library(rentrez)
library(seqinr)
library(kmer)
staphylococcus_search1 <- entrez_search(db = "nuccore", term = "(staphylococcus[ORGN] AND 16S[TITL]) NOT (genome[TITL])", retmax = 10)
staphylococcus_fetch <- entrez_fetch(db = "nuccore", id = staphylococcus_search1$ids, rettype = "fasta")
write(staphylococcus_fetch, "staphylococcus_fetch.fasta", sep = "\n")
staphylococcus_stringSet <- readDNAStringSet("staphylococcus_fetch.fasta") 
dfstaphylococcus <- data.frame(staphylococcus_Title = names(staphylococcus_stringSet), staphylococcus_Sequence = paste(staphylococcus_stringSet)) 
class(dfstaphylococcus$staphylococcus_Sequence)
is.na(dfstaphylococcus$staphylococcus_Sequence)
summary(nchar(as.character(dfstaphylococcus$staphylococcus_Sequence)))
aln_staph <- muscle::muscle(DNAStringSet(dfstaphylococcus$staphylococcus_Sequence))
staphylococcus_dis <- kdistance(x = as.DNAbin(aln_staph), k = 3)
plot(as.dendrogram(hclust(staphylococcus_dis)))
BrowseSeqs(DNAStringSet(aln_staph))
#one of my sequences seemed to be distantly related to the other sequences so I decided to Blast it. It was confirmed that it belongs to Staphylococcus aureus strain FC2929 so I kept it. As a whole, my sequences were divided into two major groups and one sequences with a distant relationship. However, I would love to compare my clusters with those determined by DNA-DNA hybridization to see if they are consistent.
